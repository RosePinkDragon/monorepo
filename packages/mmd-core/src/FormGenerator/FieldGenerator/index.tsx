import { ReactElement } from "react";
import TextInput from "./TextField";
import { TFormField } from "../../types";
import useGetFieldData from "../hooks/useGetFieldData";
import DateInputField from "./DateField";

export type TComponentRegistryProps = {
  field: any;
  value: string;
  error: string;
  touched: boolean;
  isDisabled: boolean;
};

export type TComponentRegistry = {
  [key: string]: (data: TComponentRegistryProps) => ReactElement | null;
};

export const componentRegistry: TComponentRegistry = {
  text: TextInput,
  date: DateInputField,
};

export const registerComponent = (
  type: string,
  Component: ({ field }: { field: TFormField }) => ReactElement | null
) => {
  componentRegistry[type] = Component;
};

const FieldGenerator = ({
  field,
  AfterFieldComponent,
}: {
  field: TFormField;
  AfterFieldComponent?: ({
    field,
  }: {
    field: TFormField;
  }) => ReactElement | null;
}) => {
  const Component = componentRegistry[field.type];

  const { value, error, touched, isDisabled } = useGetFieldData(field);
  if (!Component) return null;

  return (
    <>
      <Component
        field={field}
        value={value}
        error={error}
        touched={touched}
        isDisabled={isDisabled}
      />
      {AfterFieldComponent && <AfterFieldComponent field={field} />}
    </>
  );
};

export default FieldGenerator;
