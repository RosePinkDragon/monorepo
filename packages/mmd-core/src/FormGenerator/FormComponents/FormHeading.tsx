import { Typography } from "@mui/material";

const FormHeading = ({ heading }: { heading: string }) => {
  return (
    <Typography
      variant="h3"
      paddingBottom="0.8rem"
      marginBottom="1rem"
      sx={{
        position: "relative",
        ":after": {
          content: "''",
          position: "absolute",
          backgroundColor: "yellow.main",
          height: "0.2rem",
          width: "4rem",
          left: 0,
          bottom: 0,
        },
      }}
    >
      {heading}
    </Typography>
  );
};

export default FormHeading;
