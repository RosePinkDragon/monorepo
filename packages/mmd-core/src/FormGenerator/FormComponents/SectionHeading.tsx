import { Typography } from "@mui/material";

const SectionHeading = ({ heading }: { heading: string }) => {
  return (
    <Typography
      variant="h4"
      paddingBottom="0.6rem"
      marginBottom="0.4rem"
      sx={{
        position: "relative",
        ":after": {
          content: "''",
          position: "absolute",
          backgroundColor: "yellow.main",
          height: "0.2rem",
          width: "4rem",
          left: 0,
          bottom: 0,
        },
      }}
    >
      {heading}
    </Typography>
  );
};

export default SectionHeading;
