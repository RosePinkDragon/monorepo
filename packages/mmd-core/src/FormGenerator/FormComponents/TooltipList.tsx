import { Box, Typography } from "@mui/material";

const TooltipList = ({ tooltips }: { tooltips: string[] }) => {
  return (
    <Box>
      {tooltips.map((tooltip) => {
        return <Typography key={tooltip}>{tooltip}</Typography>;
      })}
    </Box>
  );
};

export default TooltipList;
