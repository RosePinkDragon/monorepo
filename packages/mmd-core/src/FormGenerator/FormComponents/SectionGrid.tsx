import { Box, Button, FormHelperText, Typography } from "@mui/material";
import Grid from "@mui/material/Unstable_Grid2";
import { FaPlus, FaTrash } from "react-icons/fa";
import { FieldArray, FormikValues, useFormikContext } from "formik";

import { TFormField } from "../../types";
import FieldGenerator from "../FieldGenerator";

interface SectiongridProps {
  name: string;
  isDisabledSection?: boolean;
  elementSize: number;
  AfterFieldComponent?: React.FC<any> | null;
  isInitialLoading?: boolean;
  isViewOnly?: boolean;
  currentFormFields: TFormField[];
}

const SectionGrid: React.FC<SectiongridProps> = ({
  name,
  isDisabledSection,
  elementSize,
  AfterFieldComponent,
  currentFormFields,
  isViewOnly,
}) => {
  const { values, errors } = useFormikContext<FormikValues>();
  return (
    <Grid
      component="fieldset"
      border="none"
      disabled={isDisabledSection}
      container
      spacing={2}
      key={name}
      py={4}
      rowSpacing={1}
      columnSpacing={{ xs: 1, sm: 2, md: 3 }}
    >
      {currentFormFields?.map((field) => {
        const fieldSize = field.colspan
          ? field.colspan * elementSize
          : elementSize;

        return (
          <Grid key={field.name} xs={12} sm={fieldSize} md={6} lg={fieldSize}>
            {field.type === "arrayField" ? (
              <FieldArray key={field.name} name={field.name}>
                {({ push, remove }) => (
                  <>
                    <Typography
                      color="primary"
                      fontWeight="bold"
                      variant="h5"
                      marginBottom={2}
                    >
                      {field.label}
                    </Typography>

                    {values[field.name]?.length > 0 &&
                      values[field.name].map((_: object, valIdx: number) => (
                        <>
                          <Box
                            key={`${field.name}_${valIdx}`}
                            display="flex"
                            justifyContent="space-between"
                            alignItems="center"
                            marginY={2}
                          >
                            <Typography>
                              {`${field.label} ${valIdx + 1}`}
                            </Typography>
                            {!isViewOnly && (
                              <Button
                                size="small"
                                type="button"
                                color="error"
                                startIcon={<FaTrash />}
                                variant="contained"
                                className="secondary"
                                onClick={() => remove(valIdx)}
                              >
                                Remove
                              </Button>
                            )}
                          </Box>
                          <Grid
                            key={field.name}
                            xs={12}
                            sm={fieldSize}
                            md={6}
                            lg={fieldSize}
                            container
                            spacing={2}
                            rowSpacing={1}
                          >
                            {field.formFields?.map((innerField) => {
                              const newName = `${field.name}.${valIdx}.${innerField.name}`;
                              return (
                                <Grid
                                  key={field.name}
                                  xs={12 / 3}
                                  sm={fieldSize / 3}
                                  md={6 / 3}
                                  lg={fieldSize / 3}
                                >
                                  <FieldGenerator
                                    field={{
                                      ...innerField,
                                      name: newName,
                                      isDisabledField: isViewOnly
                                        ? true
                                        : field.isDisabledField,
                                    }}
                                    AfterFieldComponent={
                                      AfterFieldComponent as any
                                    }
                                  />
                                </Grid>
                              );
                            })}
                          </Grid>
                        </>
                      ))}
                    {!isViewOnly && (
                      <Box
                        display="flex"
                        justifyContent="space-between"
                        alignItems="center"
                      >
                        <FormHelperText error>
                          {errors[field.name]?.toString()}
                        </FormHelperText>
                        <Button
                          sx={{
                            marginTop: "5px",
                          }}
                          size="small"
                          type="button"
                          variant="contained"
                          onClick={() => push({})}
                          startIcon={<FaPlus />}
                        >
                          Add
                        </Button>
                      </Box>
                    )}
                  </>
                )}
              </FieldArray>
            ) : (
              <FieldGenerator
                field={{
                  ...field,
                  isDisabledField: isViewOnly ? true : field.isDisabledField,
                }}
                AfterFieldComponent={AfterFieldComponent as any}
              />
            )}
          </Grid>
        );
      })}
    </Grid>
  );
};

export default SectionGrid;
