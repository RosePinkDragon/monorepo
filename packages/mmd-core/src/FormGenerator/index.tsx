import { Form, Formik, FormikTouched, FormikValues } from "formik";
import { useMemo } from "react";
import FormSection from "./FormSection";

import getInitialValuesFromForm from "../utils/getInitialValuesFromForm";
import { FormHeading } from "./FormComponents";
import { IFormGeneratorProps } from "../types/formGenerator";
import createValidationSchema from "../utils/validationSchemaGenerator";

const FormGenerator = ({
  formData,
  AfterFieldComponent,
  FormEndComponent,
  extendedSchema,
  initialFormErrors,
  initialFormValues,
  isViewOnly = false,
  submitHandler,
}: Readonly<IFormGeneratorProps>) => {
  const validationSchema = useMemo(
    () => createValidationSchema(formData, extendedSchema),
    [extendedSchema, formData]
  );

  const { sections: formSections } = formData;

  const genInitialValues = () => {
    const initialValuesFromForm = getInitialValuesFromForm(formSections);
    if (!initialFormValues) return initialValuesFromForm;
    return { ...initialFormValues, ...initialValuesFromForm };
  };

  const initialValues = genInitialValues();

  return (
    <Formik
      initialErrors={initialFormErrors}
      initialTouched={initialFormErrors as FormikTouched<FormikValues>}
      validateOnMount={Boolean(initialFormErrors)}
      initialValues={initialValues}
      validationSchema={validationSchema}
      onSubmit={(values, act) => {
        if (isViewOnly || !submitHandler) {
          return console.log(
            "you are in view only mode or you have not passed submit handler"
          );
        }
        submitHandler({ values, act });
      }}
    >
      {({ handleSubmit, values }) => (
        <Form onSubmit={handleSubmit}>
          {formData.name && <FormHeading heading={formData.name} />}
          {formSections.map((section) => {
            if (
              section.enabledOnValue &&
              section.sectionDependentOn &&
              values[section.sectionDependentOn ?? ""] !==
                section.enabledOnValue
            ) {
              return null;
            }
            return (
              <FormSection
                {...section}
                key={section.name}
                isViewOnly={isViewOnly}
                AfterFieldComponent={AfterFieldComponent}
              />
            );
          })}
          {FormEndComponent}
        </Form>
      )}
    </Formik>
  );
};

export default FormGenerator;
