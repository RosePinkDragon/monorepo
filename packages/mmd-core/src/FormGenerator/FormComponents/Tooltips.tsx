import { Icon, Tooltip } from "@mui/material";
import TooltipList from "./TooltipList";
import { AiOutlineInfoCircle } from "react-icons/ai";

const Tooltips = ({ tooltips }: { tooltips: string[] }) => {
  return (
    <Tooltip placement="top" title={<TooltipList tooltips={tooltips} />}>
      <Icon sx={{ fontSize: 20 }}>
        <AiOutlineInfoCircle />
      </Icon>
    </Tooltip>
  );
};

export default Tooltips;
