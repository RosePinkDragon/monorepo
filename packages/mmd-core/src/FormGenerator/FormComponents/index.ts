import SectionGrid from "./SectionGrid";
import FormHeading from "./FormHeading";
import SectionHeading from "./SectionHeading";

export { SectionGrid, FormHeading, SectionHeading };
