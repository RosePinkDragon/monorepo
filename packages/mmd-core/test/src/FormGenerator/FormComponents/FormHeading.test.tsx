import { render } from "@testing-library/react";
import { describe, expect, test } from "vitest";
import { SectionHeading } from "../../../../src/FormGenerator/FormComponents";

describe("renders the section heading properly", () => {
  test("renders SectionHeading with the provided heading", () => {
    const headingText = "Test Heading";
    const { getByText } = render(<SectionHeading heading={headingText} />);

    const headingElement = getByText(headingText);
    expect(headingElement).toBeInTheDocument();
  });

  test("renders SectionHeading with the provided heading and style", () => {
    const headingText = "Test Heading";
    const { getByText } = render(<SectionHeading heading={headingText} />);

    const headingElement = getByText(headingText);
    expect(headingElement).toMatchSnapshot();
    expect(headingElement).toBeInTheDocument();

    expect(headingElement).toHaveStyle({
      paddingBottom: "0.6rem",
      marginBottom: "0.4rem",
      position: "relative",
    });
  });
});
