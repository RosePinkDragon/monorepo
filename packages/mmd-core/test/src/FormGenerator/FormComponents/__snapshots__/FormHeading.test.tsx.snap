// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`renders the section heading properly > renders SectionHeading with the provided heading and style 1`] = `
<h4
  class="MuiTypography-root MuiTypography-h4 css-ruvbwe-MuiTypography-root"
>
  Test Heading
</h4>
`;
